keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,676 @@
+
+## Design Details
+
+When Kustomize processes a `Kustomization` file it uses a `ResourceAccumulator`
+(RA) object, which represents the accumulated state of the processing up to
+that point (including files, patches, transformers, variables, etc).  The RA is
+initially empty, and then Kustomize:
+
+1. Adds each of its resources (in left-to-right order) to the RA
+   (`kusttarget.accumulateResources`):
+  * Adds files to the RA directly (`accumulateFile`).
+  * Recursively processes Kustomizations, each starting with a new empty RA,
+    and with the result merged in to the RA of the parent
+    (`accumulateDirectory` and `resaccumulator.MergeAccumulator`).
+2. Adds/merges its CRDs, generators, transformers, and variables to the RA.
+  * These modifications can only be applied to entities within the RA - it is
+    an error otherwise.  This means that sibling Kustomizations are entirely
+    independent of each other.
+  * Kustomizations with the same base will cause an error when the resources
+    are accumulated, unless their GVKN is altered.
+
+`Components` are processed between step 1 and 2, and in a similar manner to
+Kustomizations, but instead of starting with an empty RA, they take the RA from
+their parent.  The component is therefore able to modify everything within the
+RA, which will contain all of the parent's resources and the result of
+processing all of the earlier components in the `components:` list. Nested
+components will work in the same way - ownership of the parent's RA will simply
+be passed down.
+
+This was implemented by changing the signature of
+`kusttarget.accumulateResources` to return a pointer to the
+`ResourceAccumulator` that the parent should use, and giving child components
+their parent's RA instead of an empty RA. This minimized the amount of code
+changed to simplify review, but a slightly more extensive refactoring would be
+recommended.
+
+Additional code changes were required to ensure that Kustomizations and normal
+files are not added to the `components:` list and Components are not added to
+`resources:`.
+
+No changes were made to the code that loads the contents of Kustomizations, so
+it should be possible to load Components from any place where one expects to
+load a Kustomization.
+
+A variation of the implementation would be to allow Components to be added to
+the `resources:` list (instead of adding them to a new `components:` list):
+
+* The basic implementation of components wouldn't need to change.
+* Unless a naming convention would be used, it would be unclear whether
+  directories in the `resources:` list are Kustomizations or Components.  A
+  naming convention could be used, but this is less explicit than using a
+  separate `components:` list. For example, store components in a
+  `components/` directory in the same way as overlay Kustomizations are often
+  stored in `overlays/`.
+
+### Test Plan
+
+(**FIXME:** Mention that we will add unit tests)
```

Comment:
This is typically an empty section for Kustomize KEPs, but we can add the following:

Add unit tests for all the points raised in the "Design details" section.

Reply:
No hard opinion, this might be redundant if we refer to testing in Graduation Criteria.